### ISSUE TITLE
Circular dependencies

### ISSUE THREAD

**Issue by pjasiun** [ckeditor/ckeditor5-design#144]
**tl;dr: `Node`, `Element`, `DocumentFragmet`, etc. can not require `Position`, `Range`, `TreeWalker`, etc.**

We had a long talk about the circular dependencies, with no clear solution unfortunately.

We have some in our project:

```
ckeditor5-engine> madge --format es6 --circular src/
treeview/node -> treeview/position -> treeview/text
treemodel/position -> treemodel/documentfragment -> treemodel/nodelist -> treemodel/characterproxy -> treemodel/node
treemodel/documentfragment -> treemodel/nodelist
treemodel/position -> treemodel/element -> treemodel/range
treemodel/element -> treemodel/range -> treemodel/treewalker
treemodel/position -> treemodel/element -> treemodel/range -> treemodel/treewalker
treemodel/operation/removeoperation -> treemodel/operation/reinsertoperation
treemodel/delta/mergedelta -> treemodel/delta/splitdelta
treemodel/delta/unwrapdelta -> treemodel/delta/wrapdelta
```

They does not cause any issue, as long there it no synchronous call needed. For instance, I introduce `Node.positionBefore` method, which created dependencies:

```
treeview/node -> treeview/position -> treeview/text
treemodel/position -> treemodel/documentfragment -> treemodel/nodelist -> treemodel/characterproxy -> treemodel/node
```

`Node` require `Position`, which require `Text`, which needs `Node` to create class, because it is a child of `Node`. But is can not load `Node` synchronously, because it is `Node` who starts a dependencies circle.

And the solution is not trivial. You can not simple add dependency to solve it. Also, the problem appears every time you require something what require `Node`, like `Element`. The solution is to require some other class from the require circle before, for instance `Position`, but then you need to remember to require `Position` every time you require `Element`, even if you do not need `Position`, what is ridiculous.

We could decrease number of dependencies avoiding `instanceof`, but using `instanceof` is useful because of two reasons:
- it checks also parent classes, so one can easily create and use child class wherever he wants instead of base class,
- it checks if it is exactly the type we need, so that it is model `Element`, not view `Element`, not DOM `Element`, but model `Element`.

We could also decrease number of dependencies avoid using `new`, by having some kind of factory, for instance `document.createElement`, but then we would have a lot of references to the `document`, what solves nothing.

Fortunately all of dependencies are always inside one directory. That is because our architecture is not that bad and we have nice layers, for instance:
1. utils,
2. model,
3. conversion,
4. controllers,
5. core,
6. features.

Modules from the next layer can require anything from the all lower layers, but not vice versa.

It looks that we have to split document "structure" and "selection" into two layers as well. It means that `Node`, `Element`, `DocumentFragmet`, etc. can not require `Position`, `Range`, `TreeWalker` etc.. We have to remove methods like `getText` (which use tree walker) or `node.getPositionBefore`. Nodes build the structure and then, on top of this structure, positions and walkers are working. It makes some logical sense.

If we keep this separation we will eliminate most of the circular dependencies, keeping only:

```
ckeditor5-engine> madge --format es6 --circular src/
treemodel/documentfragment -> treemodel/nodelist
treemodel/operation/removeoperation -> treemodel/operation/reinsertoperation
treemodel/delta/mergedelta -> treemodel/delta/splitdelta
treemodel/delta/unwrapdelta -> treemodel/delta/wrapdelta
```

First one can be solved too: `NodeList` does not need to know if something is a `DocumentFragment`. It check if given parameter is `Node`, if it's not, check if it is a `Node`s collection (iterator) and use nodes from this collection. Duck typing would work here even better then direct type checking.

Then 3 reining circular dependencies are fine and natural for me: merge <-> split, wrap <-> unwrap, remove <-> reinsert.

On the other hand we could avoid them too, by merging these files or moving `getReverse` to the factory. When we could add `madge --format es6 --circular src/` to the `git hooks` so we will not accept any code which has circular dependencies, since we spend far too much time solving them.

**Comment by pjasiun** [comment]
I decided to move this issue to the design repo, because it seems to be the right place to decide if we want to completely avoid circular dependencies and if we want to have a hook which check it for us.

**Comment by pjasiun** [comment]
Also note that all of described problems exists, because tools are imperfect. All of these dependencies could be solved, but tools would have to dynamically analyze code to understand which dependency is needed synchronously and how to resolve the circle. Unfortunately our code needs to work with these imperfect tools.

**Comment by scofalik** [comment]
It's a tough topic because it sucks that we have to re-organize and probably make our code less natural because of tools, but I guess that's the state of javascript and it always was like this.

What I don't like about merging files together is that it may be confusing which classes are in which files. Most classes have their own files but then there are some that are together. Of course this can be somehow documented (event auto-generated) but it's a mess. Also I don't like that `import` syntax changes and you have to write something like `import { UnwrapDelta } from './wrapdelta.js';`.

**Comment by Reinmar** [comment]
Great research, @pjasiun! And I like the conclusion about layering:

> It looks that we have to split document "structure" and "selection" into two layers as well. It means that Node, Element, DocumentFragmet, etc. can not require Position, Range, TreeWalker etc.. We have to remove methods like getText (which use tree walker) or node.getPositionBefore. Nodes build the structure and then, on top of this structure, positions and walkers are working. It makes some logical sense.

It's a pity though that we have to think about this at all. Our architecture really looks good now and there are just couple,  asynchronous (where dependency isn't needed synchronously), very local cycles and all of them make sense. If we had ES6 modules working natively there would be no issue at all... the only reason we have to fight this is that AMD/CJS loaders aren't able to analyse the source and find at what position they need to start loading the cycle so it all works (because it is possible to load the code now, you just need to start from the good point).

Anyway, this is the world we live in, so we need to do something and layering the code will be fine.

**Comment by pjasiun** [comment]
> What I don't like about merging files together is that it may be confusing which classes are in which files. Most classes have their own files but then there are some that are together. Of course this can be somehow documented (event auto-generated) but it's a mess. Also I don't like that `import` syntax changes and you have to write something like `import { UnwrapDelta } from './wrapdelta.js';`. 

I did some additional research during the weekend, mostly reading SO discussions, and I realised 2 thinks:

First, working in dynamic typing language we live comfortable environment. Because we do not need to import module to declare variable, most of our circular dependencies are hidden. When we check the type using `instanceof` we do not create a new dependency, but reveal one which already was there.

Second, most opinions agree that circular dependencies are fine as long they are "local", on the same level. This is something we can  not avoid and if we will try we will create even bigger problems. So I agree that we should keep merge <-> split, wrap <-> unwrap and remove <-> reinsert dependencies, especially when all of them are asynchronous. Harmful circular dependencies appear when you have cross modules cycles, but this is something we managed to avoid creating well defined abstraction layers.

**Comment by Reinmar** [comment]
We're fine at the moment and we don't encounter issues anymore. We still have small circular deps, but in a well-controlled part of the engine (and nowhere else).
